### PR TITLE
Apply Shim to Cumulative Death s.t. Latest Observed Equals Latest Model

### PR DESCRIPTION
**Pull request by BrettBoval**
This code shims the cumulative deaths timeseries in the API so that the cumulative deaths from the model agrees with the cumulative deaths from outside observed sources.

This PR addresses issue https://trello.com/c/ZDNt3or6/247-shim-model-cumulative-deaths-to-observed.

### Please Check
- [ ] Are the [JSON schemas](https://github.com/covid-projections/covid-data-model/tree/master/api/schemas) updated?
- [ ] Are the [api docs](https://github.com/covid-projections/covid-data-model/blob/master/api/README.V1.md) updated?
- [ ] Are tests passing?

### CODE REVIEW

**Comment by BrettBoval** [comment]
@TomGoBravo can you take a look. The main thing is a feature that shims the deaths timeseries to make sure that observed today equals model today. This was a business requirement that surfaced Wednesday. 

In the course of fixing it, I've changed some of the more verbose warnings that are crowding stdout and that no one is using. I've kept the warnings, but not the dataframe dumps.

Finally, there were two changes you recommended on the last PR that slipped through. They've been added here.

**Comment by BrettBoval** [comment]
And please recommend the most obvious 1-3 test cases for this. I know I should put them in, but I'm too tired at the moment.

File: pyseir/deployment/webui_data_adaptor_v1.py [hunk 1]
```diff
@@ -393,9 +418,10 @@ def generate_state(self, whitelisted_county_fips: list, states_only=False):
 
 
 if __name__ == "__main__":
+    dataset_cache.set_pickle_cache_dir()
```
Comment (mikelehen) [inline pyseir/deployment/webui_data_adaptor_v1.py:421]: Is this intentional?  Not sure if/how it relates to these changes?

File: pyseir/inference/infer_rt.py [hunk 2]
```diff
@@ -13,12 +13,6 @@
 from pyseir.utils import AggregationLevel, TimeseriesType
 from pyseir.utils import get_run_artifact_path, RunArtifact
 from pyseir.parameters.parameter_ensemble_generator import ParameterEnsembleGenerator
-from structlog.threadlocal import bind_threadlocal, clear_threadlocal, merge_threadlocal
-from structlog import configure
-from enum import Enum
-
-configure(processors=[merge_threadlocal, structlog.processors.KeyValueRenderer()])
-log = structlog.get_logger(__name__)
```
Comment (mikelehen) [inline pyseir/inference/infer_rt.py:21]: Just flagging that I have no context on these changes.  So if you think they warrant further review, you should ping somebody else or we can chat through them.  If they're low-risk / already-discussed-with-folks / whatever, then cool.

Comment (BrettBoval) [reply pyseir/inference/infer_rt.py:21]: Yeah. These were already discussed promises I made to Tom last week and that he signed off on. I added in that code last week as a patch and how now cleaned it up.

File: pyseir/deployment/webui_data_adaptor_v1.py [hunk 3]
```diff
@@ -211,6 +212,24 @@ def map_fips(self, fips: str) -> None:
             t0_simulation=t0_simulation, fips=fips, pyseir_outputs=pyseir_outputs,
         )
 
+        # Here is where we shimmy just for cumulative death right now.
+        # We will shim all suppression policies by the same amount (since historical tracking error
+        # for all policies is the same).
+        baseline_policy = "suppression_policy__inferred"  # This could be any valid policy
+        model_death_ts = pyseir_outputs[baseline_policy]["total_deaths"]["ci_50"]
```
[the remark this answers is not in the document]
Comment (BrettBoval) [reply pyseir/deployment/webui_data_adaptor_v1.py:219]: Correct. It's misleading, but that's the naming convention throughout the rest of the code.

File: pyseir/deployment/webui_data_adaptor_v1.py [hunk 4]
```diff
@@ -211,6 +212,24 @@ def map_fips(self, fips: str) -> None:
             t0_simulation=t0_simulation, fips=fips, pyseir_outputs=pyseir_outputs,
         )
 
+        # Here is where we shimmy just for cumulative death right now.
```
[the remark this answers is not in the document]
Comment (BrettBoval) [reply pyseir/deployment/webui_data_adaptor_v1.py:215]: Agreed and resolved.

File: pyseir/deployment/model_to_observed_shim.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    model_latest_cum_deaths = model_death_ts[idx]
+
+    if np.isnan(observed_latest_cum_deaths):
+        death_shim = 0
+    elif observed_latest_cum_deaths == 0:
+        death_shim = 0
+    else:
+        # To be consistent with the other shims coming, I will shim all values by the same amount
+        # so that the cumulative model today equals cumulative observed today. This will shift the
+        # future values by the same amount.
```
[the remark this answers is not in the document]
Comment (BrettBoval) [reply pyseir/deployment/model_to_observed_shim.py:43]: This patch originally had the switch over to shim for hospitalized, deaths, and ICU. I've removed those others for now, but this is an artifact of that. I'll remove it.

File: pyseir/deployment/model_to_observed_shim.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+from typing import Sequence
+
+import numpy as np
+
+
+def shim_deaths_model_to_observations(
+    model_death_ts: Sequence, idx: int, observed_latest: dict, log
+):
+    """Enforce the constraint that cumulative model deaths up to today equals cumulative observed
+    deaths up to today. Take model outputs and shim them s.t. the latest observed value matches the
```
[the remark this answers is not in the document]
Comment (BrettBoval) [reply pyseir/deployment/model_to_observed_shim.py:10]: done.

File: pyseir/deployment/model_to_observed_shim.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    same value for the outputted model.
+
+    Parameters
+    ----------
+    model_death_ts
+        Model array sequence for cumulative deaths
+    idx
+        Index on which to align
+    observed_latest
+        Dictionary of latest values taken from combined dataset for the fips of interest.
```
[the remark this answers is not in the document]
Comment (BrettBoval) [reply pyseir/deployment/model_to_observed_shim.py:20]: Agreed. This module originally had shims for multiple timeseries and more structured routing. That was lost when I removed the hospitalized and icu shims. I will refactor this when I add those back in.

File: pyseir/deployment/model_to_observed_shim.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    # There are inconsistent None/"NaN" -> force all to np.nan for this scope.
+    if observed_latest_cum_deaths is None:
+        observed_latest_cum_deaths = np.nan
+
+    model_latest_cum_deaths = model_death_ts[idx]
+
+    if np.isnan(observed_latest_cum_deaths):
+        death_shim = 0
+    elif observed_latest_cum_deaths == 0:
```
[the remark this answers is not in the document]
Comment (BrettBoval) [reply pyseir/deployment/model_to_observed_shim.py:38]: comment added.

[the remark this answers is not in the document]
Comment (tashwoods) [reply pyseir/deployment/model_to_observed_shim.py:39]: So with this change if the last value is zero, we apply no-correction and also use that value? If so, that seems wrong. I would prefer to use the last non-zero value unless there have been 3 days of zero values or so.

File: pyseir/deployment/webui_data_adaptor_v1.py [hunk 9]
```diff
@@ -393,9 +417,10 @@ def generate_state(self, whitelisted_county_fips: list, states_only=False):
 
 
 if __name__ == "__main__":
+    dataset_cache.set_pickle_cache_dir()
     # Need to have a whitelist pre-generated
     # Need to have state output already built
     mapper = WebUIDataAdaptorV1(
-        state="California", output_interval_days=4, run_mode="can-inference-derived"
+        state="California", output_interval_days=1, run_mode="can-inference-derived"
```
Comment (tashwoods) [inline pyseir/deployment/webui_data_adaptor_v1.py:424]: Why did you change this?

File: pyseir/inference/infer_rt.py [hunk 10]
```diff
@@ -225,7 +219,7 @@ def apply_gaussian_smoothing(self, timeseries_type, plot=False, smoothed_max_thr
             This parameter allows you to filter out entire series
             (e.g. NEW_DEATHS) when they do not contain high enough
             numeric values. This has been added to account for low-level
-            constant smoothed values having a dispropotionate effect on
+            constant smoothed values having a disproportionate effect on
```
Comment (tashwoods) [inline pyseir/inference/infer_rt.py:222]: nice catch haha : )

File: pyseir/inference/infer_rt.py [hunk 11]
```diff
@@ -242,7 +236,7 @@ def apply_gaussian_smoothing(self, timeseries_type, plot=False, smoothed_max_thr
         """
         timeseries_type = TimeseriesType(timeseries_type)
         dates, times, timeseries = self.get_timeseries(timeseries_type)
-        bind_threadlocal(timeseries_type=timeseries_type.value)
+        self.log = self.log.bind(timeseries_type=timeseries_type.value)
```
Comment (tashwoods) [inline pyseir/inference/infer_rt.py:239]: Are these changes already in Alex's code? If so, I prefer we just merge his branch, rather than having the infer_rt.py changes across multiple PRs.

File: pyseir/deployment/webui_data_adaptor_v1.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tashwoods) [reply pyseir/deployment/webui_data_adaptor_v1.py:218]: So here you are returning the upper value of the 50% CI for the death time series?

File: pyseir/deployment/webui_data_adaptor_v1.py (hunk 9 above)
Comment (BrettBoval) [reply pyseir/deployment/webui_data_adaptor_v1.py:424]: It's always overwritten when we instantiate this class, and we use 1 in the api.

File: pyseir/inference/infer_rt.py (hunk 11 above)
Comment (BrettBoval) [reply pyseir/inference/infer_rt.py:239]: These are different changes. I'll be happy to work through the merge conflicts if they show up.

File: pyseir/deployment/model_to_observed_shim.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (BrettBoval) [reply pyseir/deployment/model_to_observed_shim.py:39]: This is a cumulative metric, so we don't have the worries the same way that we do of daily deaths going to zero.

The fix for this is to tighten up validation in combined_datasets.build_us_latest_with_all_fields. Right now I don't trust the zeros in cumulative deaths given we are months into this, so I ignore them and don't scale. I don't want to write a data validator in every feature. I'm working with Tom to have this be upstream. Once that's there, then I'll remove this.

File: pyseir/deployment/webui_data_adaptor_v1.py (hunk 9 above)
Comment (BrettBoval) [reply pyseir/deployment/webui_data_adaptor_v1.py:424]: It's also easier to reason about the spacing of the timeseries lengths with a day=1 spacing. If someone wants to revert back to 4 day spacing they can make it with an explicit call.

**Comment by BrettBoval** [comment]
Response:
"how you are handling timeseries when the last entry is zero"
- I get the latest observed value for cumulative deaths for a given fips. If that is zero or None or np.NaN I don't apply a shim. I do not have access to the timeseries of observed via the accessor I want to call. See the comment above for more information.

"what model_death_ts is returning and how it is being used downstream"
- model_death_ts is returning anything in this case in the way I think you're asking. It's the model's estimate for cumulative deaths as a function of time. The FE looks at that timeseries and takes the today+30(use to be 90) value, rounds it to the 1000s and then presents it in the deaths timeseries.

File: pyseir/deployment/model_to_observed_shim.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
+from typing import Sequence
+
+import numpy as np
+
+
+def shim_deaths_model_to_observations(
```
Comment (ghop02) [inline pyseir/deployment/model_to_observed_shim.py:6]: Small comment from the PR grave - it would be great to throw in a small unit test for this function.

Going to be looking for opportunities to add slightly more testing to make refactoring easier in the future/examples of how the current functions are used.
